### PR TITLE
Use ToString in Interpolation

### PR DESCRIPTION
**Pull request by magnus-madsen**
- [x] Simplify weeding.
- [x] Add source positions to interpolations.
- [x] Add unit tests.
- [x] Rebase onto master.

### CODE REVIEW

File: main/test/flix/Test.Exp.Interpolation.flix [hunk 1]
```diff
@@ -120,3 +215,60 @@ namespace Test/Exp/Interpolation {
         s.toUpperCase() as & Pure
 
 }
+
+namespace ToString {
+
+    pub class ToString[a] {
+        pub def toString(x: a): String
+    }
```
Comment (mlutze) [inline main/test/flix/Test.Exp.Interpolation.flix:223]: Why isn't this causing an error for clashing with the existing ToString class?

File: main/src/ca/uwaterloo/flix/language/phase/Weeder.scala [hunk 2]
```diff
@@ -1016,39 +1014,44 @@ object Weeder extends Phase[ParsedAst.Program, WeededAst.Program] {
...
-          // Case 3: Return the string if there is only a single string.
-          WeededAst.Expression.Str(s, loc).toSuccess
+      val loc = mkSL(sp1, sp2)
 
-        case ps =>
-          // Case 4: Construct a sequence of string append expressions.
+      parts match {
+        case Seq(ParsedAst.InterpolationPart.StrPart(innerSp1, lit, innerSp2)) =>
+          // Special case: We have a constant string. Simply return it.
+          WeededAst.Expression.Str(lit, mkSL(innerSp1, innerSp2)).toSuccess
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder.scala:1039]: Does this indicate a likely bug that we should report?

File: main/test/flix/Test.Exp.Interpolation.flix [hunk 3]
```diff
@@ -111,6 +111,101 @@ namespace Test/Exp/Interpolation {
         let x = "a";
         "$${x}$" == "$a$"
 
+    @test
+    def interpolation21(): Bool =
+        let x = true;
+        "${x}" == "true"
```
Comment (mlutze) [inline main/test/flix/Test.Exp.Interpolation.flix:117]: General question. Scala and Python (and probably others) have a prefix to mark that a string is interpolated, which lets a programmer explicitly indicate their intention to use interpolation. What influenced the choice not to use a prefix?

File: main/src/ca/uwaterloo/flix/language/phase/Weeder.scala [hunk 4]
```diff
@@ -1016,39 +1014,44 @@ object Weeder extends Phase[ParsedAst.Program, WeededAst.Program] {
...
+        WeededAst.Expression.Binary(op, e1, e2, loc)
+      }
 
-        case Seq(ParsedAst.InterpolationPart.ExpPart(e)) =>
-          // Case 2: Return the expression if there is only a single expression.
-          visitExp(e)
+      /**
+        * Returns an expression that applies `toString` to the result of the given expression `e`.
+        */
+      def mkApplyShow(e: WeededAst.Expression, sp1: SourcePosition, sp2: SourcePosition): WeededAst.Expression = {
```
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Weeder.scala:1029]: Fixed

File: main/src/ca/uwaterloo/flix/language/phase/Weeder.scala (hunk 2 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Weeder.scala:1039]: No, this is just an artifact of the fact that `"abc"` is a string interpolator without anything in it.

File: main/test/flix/Test.Exp.Interpolation.flix (hunk 3 above)
Comment (magnus-madsen) [reply main/test/flix/Test.Exp.Interpolation.flix:117]: Good question: I have no answer. I was thinking about this yesterday.

it seems to me that Python/Scala etc. all added interpolation later hence they were forced to come up with a new type of string. But for us, why not have all strings be interpolation? I see very little downside. If you want to write the string "${x}$ you can also just do "$" + "{x}". (or we could add escapes).

So, should all strings be interpolated?

File: main/test/flix/Test.Exp.Interpolation.flix (hunk 1 above)
Comment (magnus-madsen) [reply main/test/flix/Test.Exp.Interpolation.flix:223]: Ah-

We are slowly moving towards a situation where the compiler tests do not depend on the standard library. (This way it is much faster to run all tests). Most tests now are independent files that simply declare any constructs that they may need.

Comment (magnus-madsen) [reply main/test/flix/Test.Exp.Interpolation.flix:223]: Good question!

File: main/test/flix/Test.Exp.Interpolation.flix (hunk 3 above)
Comment (mlutze) [reply main/test/flix/Test.Exp.Interpolation.flix:117]: I think that's fair enough. This will be even less of an issue when we get raw strings.

**Pull request merged**